addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,448 @@
+	packages := append([]*runtanzuv1alpha3.ClusterBootstrapPackage{
+		clusterBootstrap.Spec.CNI,
+		clusterBootstrap.Spec.CPI,
+		clusterBootstrap.Spec.CSI,
+	}, clusterBootstrap.Spec.AdditionalPackages...)
+
+	var cbStatusChanged bool
+
+	for _, pkg := range packages {
+		if pkg == nil {
+			continue
+		}
+		cbConditionChanged, err := r.reconcileClusterBootstrapStatus(ctx, clusterClient, &clusterBootstrap.Status, cluster.Name, clusterRole, pkg.RefName, constants.TKGSystemNS, log)
+		cbStatusChanged = cbStatusChanged || cbConditionChanged
+		if err != nil {
+			// in this case, we delete the existing condition corresponding to the package in the ClusterBootstrapStatus (if existing); as the corresponding pkgi or package resources do not exist for the package anymore
+			// then just continue with collecting PackageInstallStatus for other packages
+			cbStatusChanged = cbStatusChanged || r.removeConditionIfExistsForPkgName(&clusterBootstrap.Status, pkg.RefName)
+		}
+	}
+
+	// Note: kapp ctrl pkgi exists only for the workload cluster
+	// it is installed under cluster.Namespace in the management cluster and should be handled separately
+	if clusterRole == clusterRoleWorkload && clusterBootstrap.Spec.Kapp != nil {
+		cbConditionChanged, err := r.reconcileClusterBootstrapStatus(ctx, r.Client, &clusterBootstrap.Status, cluster.Name, clusterRole, clusterBootstrap.Spec.Kapp.RefName, cluster.Namespace, log)
+		cbStatusChanged = cbStatusChanged || cbConditionChanged
+		if err != nil {
+			// in this case, we delete the existing condition corresponding to the package in the ClusterBootstrapStatus (if existing); as the corresponding pkgi or package resources do not exist for the package anymore
+			// then just proceed with updating the ClusterBootstrapStatus for other packages in one update operation
+			cbStatusChanged = cbStatusChanged || r.removeConditionIfExistsForPkgName(&clusterBootstrap.Status, clusterBootstrap.Spec.Kapp.RefName)
+		}
+	}
+
+	if !cbStatusChanged {
+		log.Info("no change in the reconciliation status of any of the core/additional packages in the cluster")
+		return nil
+	}
+
+	if err := r.Client.Status().Update(ctx, clusterBootstrap); err != nil {
+		return err
+	}
+
+	log.Info("Successfully updated ClusterBootstrapStatus")
+	return nil
+}
+
+// reconcileClusterBootstrapStatus reconciles clusterBootstrapStatus by setting conditions corresponding to all core/additional packages
+// The Status update happens in the caller function
+func (r *PackageInstallStatusReconciler) reconcileClusterBootstrapStatus(
+	ctx context.Context,
+	clusterClient client.Client,
+	clusterBootstrapStatus *runtanzuv1alpha3.ClusterBootstrapStatus,
+	clusterName string,
+	clusterRole ClusterRole,
+	pkgName string,
+	pkgiNamespace string,
+	log logr.Logger) (bool, error) {
+
+	var pkgiName, pkgShortname string
+
+	c := clusterClient
+	// controller-runtime's client (r.Client) is not able to find aggregated api server resources.
```

Comment:
I'm a bit confused. Aren't those all controller-runtime clients? Or is it rather about that we have to use a controller-runtime uncached client for that?

If yes, don't we have a problem with doing this on a workload cluster as it is using a client with cache retrieved from the ClusterCacheTracker?

Reply:
We were seeing that the controller-runtime cached client used by manager would return not found for a GET on an aggregated API server resource but list would work fine so we use [uncached client](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/client.go#L75) specifically for fetching PackageMetadata resources. cc @danniel1205

Reply:
@sbueringer you are right that both are indeed controller runtime clients and the difference is only on one might use the cache to read and the other one not (I should improve my comment).

I only use this on management cluster though. For the following check, for the second conditional check, kapp controller is still installed inside the management cluster but in workload cluster's namespace. I never use ```r.aggregatedAPIResourcesClient``` for workload clusters.
```
if clusterRole == clusterRoleManagement || strings.Contains(pkgName, kappCtrlPkgPrefix) {..}
```

Reply:
Okay sounds good. I think the missing piece for me is that we have an aggregated API server in the management cluster and not in the workload cluster. Is that correct? 

I really don't know a lot about how Tanzu details like this :)

Reply:
Yes it seems like, I think cause the way resources like Package get available in the management & workload clusters are different; and that's why we don't have such problems in the workload cluster.  @danniel1205 Please correct me if wrong on this raesoning?